src/condor_starter.V6.1/filter.cpp

```diff
@@ -0,0 +1,181 @@
+#include "condor_common.h"
+#include "condor_debug.h"
+
+#include "classad/classad.h"
+#include "filter.h"
+#include "compat_classad.h"
+
+#include <map>
+#include <string>
+#include <vector>
+#if defined(CXX_TWENTY_IS_REAL_TO_ME)
+#include <algorithm>
+#include <ranges>
+#endif   /* CXX_TWENTY_IS_REAL_TO_ME */
+
+//
+// Excepting the special case, these two function are identical, so
+// they should be refactored together.  Maybe by having a callback
+// for each CLASSAD_VALUE (with the extracted ClassAd and holding
+// attribute name as arguments)?
+//
+// ... would probably be less ugly as a two-pass algorithm in the
+// top-level wrapper.
+//
+
+
+// The caller owns (is responsible for delete'ing) the returned ClassAd.
+classad::ClassAd *
+filterTransferErrorData( const classad::ClassAd & ad ) {
+    using namespace classad;
+
+    std::map<std::string, classad::Value::ValueType, CaseIgnLTStr> LegalAttributes {
+        { "DeveloperData",                  Value::CLASSAD_VALUE },
+        { "IntermediateServerErrorType",    Value::STRING_VALUE },
+        { "IntermediateServer",             Value::STRING_VALUE },
+        { "ErrorType",                      Value::STRING_VALUE },
+        { "PluginVersion",                  Value::STRING_VALUE },
+        { "PluginLaunched",                 Value::BOOLEAN_VALUE },
+        { "FailedName",                     Value::STRING_VALUE },
+        { "FailureType",                    Value::STRING_VALUE },
+        { "FailedServer",                   Value::STRING_VALUE },
+        { "ShouldRefresh" ,                 Value::BOOLEAN_VALUE },
+        { "ErrorCode",                      Value::INTEGER_VALUE },
+        { "ErrorSrring",                    Value::STRING_VALUE },
+    };
+
+
+    // The ClassAd iterator isn't sorted, which set_intersection() requires.
+    classad::References attributes;
+    sGetAdAttrs( attributes, ad );
+
+    std::vector<std::string> intersection;
+#if defined(CXX_TWENTY_IS_REAL_TO_ME)
+    std::ranges::set_intersection(
+        attributes,
+        LegalAttributes | std::ranges::views::keys,
+        std::back_inserter( intersection ),
+        CaseIgnLTStr()
+    );
+#else
+    for( const auto & attribute : attributes ) {
+        if( LegalAttributes.count(attribute) != 0 ) {
+            intersection.push_back( attribute );
+        }
+    }
+#endif
+
+
+    if( intersection.empty() ) { return NULL; }
+
+    classad::ClassAd * filteredAd = new classad::ClassAd();
+    for( const auto & attribute : intersection ) {
+        classad::Value v;
+        ExprTree * e = ad.Lookup( attribute );
+        if( e->Evaluate(v) ) {
```

Comment:
This is not necessarily safe in the general case when the expression `e` is a function that returns an ad and you subsequently copy the value into a new ad.  It should be safe for all cases we care about, but it would be safer to just examine the expression here and evaluate it only if it is not a classad, and accept only POD types for expressions that you evaluate.  (or never evaluate, and accept only ads and literals here...)